### PR TITLE
*: fix ctrl+c can not kill tidb during statistics init

### LINKED ISSUE #54589: Ctrl+C can't make tidb exit with 1M tables using infoschema v2
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

I'm using @@global.tidb_schema_cache_size = '512MB' and there are 1M tables.
Also the configuration `force-init-stats = false`
Start TiDB and wait 4000 port ready.
Then Ctrl+C to exit tidb

### 2. What did you expect to see? (Required)

TiDB exit

### 3. What did you see instead (Required)

It takes 12m54s to exit!

```
[2024/07/11 16:25:27.006 +08:00] [INFO] [domain.go:1164] ["domain closed"] ["take time"=12m54.727120057s]
[2024/07/11 16:25:27.006 +08:00] [INFO] [gc_worker.go:223] [quit] [category="gc worker"] [uuid=642836439700005]
```

pd service discovery and tso dispatcher exit after domain close
```
[2024/07/11 16:25:27.006 +08:00] [INFO] [pd_service_discovery.go:550] ["[pd] exit member loop due to context canceled"]
[2024/07/11 16:25:27.007 +08:00] [INFO] [tso_dispatcher.go:210] ["[tso] stop fetching the pending tso requests due to context canceled"] [dc-location=global]
[2024/07/11 16:25:27.007 +08:00] [INFO] [tso_client.go:147] ["[tso] exit tso dispatcher check loop"]
[2024/07/11 16:25:27.007 +08:00] [INFO] [tso_dispatcher.go:350] ["[tso] exit tso connection contexts updater"] [dc-location=global]
[2024/07/11 16:25:27.007 +08:00] [INFO] [tso_dispatcher.go:171] ["[tso] exit tso dispatcher"] [dc-location=global]
[2024/07/11 16:25:27.007 +08:00] [INFO] [tso_batch_controller.go:160] ["[pd] clear the tso batch controller"] [max-batch-size=10000] [best-batch-size=1] [collected-request-count=0] [pending-request-count=0]
[2024/07/11 16:25:27.006 +08:00] [INFO] [tso_dispatcher.go:137] ["[tso] exit tso deadline watcher"] [dc-location=global]
[2024/07/11 16:25:27.007 +08:00] [INFO] [resource_manager_client.go:290] ["[resource manager] exit resource token dispatcher"]
[2024/07/11 16:25:27.007 +08:00] [INFO] [tso_client.go:157] ["[tso] closing tso client"]
[2024/07/11 16:25:27.007 +08:00] [INFO] [tso_client.go:162] ["[tso] close tso client"]
[2024/07/11 16:25:27.007 +08:00] [INFO] [tso_batch_controller.go:160] ["[pd] clear the tso batch controller"] [max-batch-size=10000] [best-batch-size=1] [collected-request-count=0] [pending-request-count=0]
[2024/07/11 16:25:27.007 +08:00] [INFO] [tso_client.go:164] ["[tso] tso client is closed"]
[2024/07/11 16:25:27.007 +08:00] [INFO] [pd_service_discovery.go:637] ["[pd] close pd service discovery client"]
[2024/07/11 16:25:27.007 +08:00] [INFO] [client.go:347] ["[pd] http client closed"] [source=tikv-driver]
```

sql cpu collector and cgroup monitor exit after domain:
```
[2024/07/11 16:25:27.007 +08:00] [INFO] [cpu.go:98] ["sql cpu collector stopped"]
[2024/07/11 16:25:27.007 +08:00] [INFO] [cgmon.go:77] ["cgroup monitor stopped"]
[2024/07/11 16:25:27.007 +08:00] [INFO] [cpuprofile.go:130] ["parallel cpu profiler stopped"]
```

gc worker print such log every minute, it's not exited after receiving the signal:
```
[2024/07/11 16:17:42.998 +08:00] [WARN] [gc_worker.go:274] ["check leader"] [category="gc worker"] [error="[domain:8027]Information sche
[2024/07/11 16:18:42.995 +08:00] [WARN] [gc_worker.go:274] ["check leader"] [category="gc worker"] [error="[domain:8027]Information sche
[2024/07/11 16:19:42.999 +08:00] [WARN] [gc_worker.go:274] ["check leader"] [category="gc worker"] [error="[domain:8027]Information sche
[2024/07/11 16:20:42.999 +08:00] [WARN] [gc_worker.go:274] ["check leader"] [category="gc worker"] [error="[domain:8027]Information sche
[2024/07/11 16:21:42.994 +08:00] [WARN] [gc_worker.go:274] ["check leader"] [category="gc worker"] [error="[domain:8027]Information sche
[2024/07/11 16:22:42.999 +08:00] [WARN] [gc_worker.go:274] ["check leader"] [category="gc worker"] [error="[domain:8027]Information sche
[2024/07/11 16:23:42.999 +08:00] [WARN] [gc_worker.go:274] ["check leader"] [category="gc worker"] [error="[domain:8027]Information sche
[2024/07/11 16:24:42.999 +08:00] [WARN] [gc_worker.go:274] ["check leader"] [category="gc worker"] [error="[domain:8027]Information sche
```

The update stats worker exited immediately:
```
[2024/07/11 16:12:34.965 +08:00] [INFO] [domain.go:2490] ["updateStatsWorker exit preprocessing finished"]
[2024/07/11 16:12:34.965 +08:00] [INFO] [domain.go:2533] ["updateStatsWorker exited."]
```

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #54589

Problem Summary:

When there a a lot of tables, statistics initialize takes a long time.
For example, 1M tables takes 13-14min with infoschema v2 currently.

**When TiDB start up and initialize statistics, it does not handle any interrupt signal.**
So if I start tidb and then Ctrl+C, domain.Close will wait the background worker to exit.
But the statistics load worker are handling the statistics initialization, it never response to the exit signal.
So tidb can't exit.

### What changed and how does it work?

Add a context.Context to the related statistics code and handle context.Cancel signal.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)

Before this commit, the log is this:

```
[2024/07/11 16:25:27.006 +08:00] [INFO] [domain.go:1164] ["domain closed"] ["take time"=12m54.727120057s]
```

And after the fix, tidb can exit quickly:

```
[2024/07/12 10:42:00.562 +08:00] [INFO] [wait_group_wrapper.go:140] ["background process exited"] [source=domain] [process=loadStatsWorker]
[2024/07/12 10:42:02.552 +08:00] [INFO] [manager.go:326] ["revoke session"] ["owner info"="[bindinfo] /tidb/bindinfo/owner ownerManager 906be1c7-194e-4bb4-a742-09a9a4db15de"] [error="rpc error: code = Canceled desc = grpc: the client connection is closing"]
[2024/07/12 10:42:02.552 +08:00] [INFO] [domain.go:1984] ["globalBindHandleWorkerLoop exited."]
[2024/07/12 10:42:02.552 +08:00] [INFO] [wait_group_wrapper.go:140] ["background process exited"] [source=domain] [process=globalBindHandleWorkerLoop]
[2024/07/12 10:42:02.574 +08:00] [INFO] [manager.go:326] ["revoke session"] ["owner info"="[stats] /tidb/stats/owner ownerManager 906be1c7-194e-4bb4-a742-09a9a4db15de"] [error="rpc error: code = Canceled desc = grpc: the client connection is closing"]
[2024/07/12 10:42:02.574 +08:00] [INFO] [domain.go:2509] ["updateStatsWorker exit preprocessing finished"]
[2024/07/12 10:42:02.574 +08:00] [INFO] [domain.go:2552] ["updateStatsWorker exited."]
[2024/07/12 10:42:02.574 +08:00] [INFO] [wait_group_wrapper.go:140] ["background process exited"] [source=domain] [process=updateStatsWorker]

[2024/07/12 10:42:02.575 +08:00] [INFO] [domain.go:1168] ["domain closed"] ["take time"=2.55484294s]

[2024/07/12 10:42:02.575 +08:00] [INFO] [gc_worker.go:223] [quit] [category="gc worker"] [uuid=6429345da3c0005]
[2024/07/12 10:42:02.575 +08:00] [INFO] [pd_service_discovery.go:550] ["[pd] exit member loop due to context canceled"]
[2024/07/12 10:42:02.575 +08:00] [INFO] [tso_client.go:147] ["[tso] exit tso dispatcher check loop"]
[2024/07/12 10:42:02.575 +08:00] [INFO] [tso_dispatcher.go:137] ["[tso] exit tso deadline watcher"] [dc-location=global]
[2024/07/12 10:42:02.575 +08:00] [INFO] [tso_dispatcher.go:210] ["[tso] stop fetching the pending tso requests due to context canceled"] [dc-location=global]
[2024/07/12 10:42:02.575 +08:00] [INFO] [tso_dispatcher.go:350] ["[tso] exit tso connection contexts updater"] [dc-location=global]
[2024/07/12 10:42:02.575 +08:00] [INFO] [tso_dispatcher.go:171] ["[tso] exit tso dispatcher"] [dc-location=global]
[2024/07/12 10:42:02.575 +08:00] [INFO] [tso_batch_controller.go:160] ["[pd] clear the tso batch controller"] [max-batch-size=10000] [best-batch-size=1] [collected-request-count=0] [pending-request-count=0]
[2024/07/12 10:42:02.575 +08:00] [INFO] [resource_manager_client.go:290] ["[resource manager] exit resource token dispatcher"]
[2024/07/12 10:42:02.575 +08:00] [INFO] [tso_client.go:157] ["[tso] closing tso client"]
[2024/07/12 10:42:02.575 +08:00] [INFO] [tso_client.go:162] ["[tso] close tso client"]
[2024/07/12 10:42:02.575 +08:00] [INFO] [tso_batch_controller.go:160] ["[pd] clear the tso batch controller"] [max-batch-size=10000] [best-batch-size=1] [collected-request-count=0] [pending-request-count=0]
```

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by hawkingrei — APPROVED** [review]

File: pkg/statistics/handle/bootstrap.go [hunk 1]
```diff
@@ -53,10 +53,19 @@ type MaxTidRecord struct {
...
-func (h *Handle) initStatsMeta4Chunk(is infoschema.InfoSchema, cache statstypes.StatsCache, iter *chunk.Iterator4Chunk) {
+func (h *Handle) initStatsMeta4Chunk(ctx context.Context, is infoschema.InfoSchema, cache statstypes.StatsCache, iter *chunk.Iterator4Chunk) {
 	var physicalID, maxPhysicalID int64
 	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
 		physicalID = row.GetInt64(1)
+
+		// Detect the context cancel signal, since it may take a long time for the loop.
+		// TODO: add context to TableInfoByID and remove this code block?
+		select {
+		case <-ctx.Done():
```
Comment (lance6716) [inline pkg/statistics/handle/bootstrap.go:64]: if ctx.Err() != nil { return }

File: pkg/statistics/handle/cache/statscache.go [hunk 2]
```diff
@@ -89,6 +90,15 @@ func (s *StatsCacheImpl) Update(is infoschema.InfoSchema) error {
 		physicalID := row.GetInt64(1)
 		modifyCount := row.GetInt64(2)
 		count := row.GetInt64(3)
+
+		// Detect the context cancel signal, since it may take a long time for the loop.
+		// TODO: add context to TableInfoByID and remove this code block?
+		select {
```
Comment (lance6716) [inline pkg/statistics/handle/cache/statscache.go:96]: ditto

**Review by lance6716 — APPROVED** [review]
rest lgtm

File: pkg/statistics/handle/bootstrap.go (hunk 1 above)
Comment (tiancaiamao) [reply pkg/statistics/handle/bootstrap.go:64]: done

Comment (lance6716) [reply pkg/statistics/handle/bootstrap.go:64]: No I mean replace the whole select {...}.

Comment (zimulala) [reply pkg/statistics/handle/bootstrap.go:64]: I think @lance6716 means we can use `if ctx.Err() != nil { return }` instead of line63-69.

**Review by zimulala — APPROVED** [review]

**Pull request merged**